internal/trie/node/copy.go

```diff
@@ -3,76 +3,136 @@
 
 package node
 
+// DefaultCopySettings returns the following copy settings:
+// - children are NOT deep copied recursively
+// - the HashDigest field is left empty on the copy
+// - the Encoding field is left empty on the copy
+// - the key field is deep copied
+// - the value field is deep copied
+func DefaultCopySettings() CopySettings {
+	return CopySettings{
+		CopyKey:   true,
+		CopyValue: true,
+	}
+}
```

Comment:
To have it more 'immutable' / less 'global variable'.
This allows to easily get the default copy settings and then modify it, without affecting other parts of the code.

Reply:
I think it's understood that you shouldn't be modifying package level variables.  Even if someone decides to copy and modify a few attributes, it won't update the package level instance.  I don't see the need to call a function for this and allocating more memory.

Reply:
> I think it's understood that you shouldn't be modifying package level variables.

Disagree :smile_cat: So many times I have seen (I guess even myself) inadvertently modifying a global variable.

> Even if someone decides to copy and modify a few attributes, it won't update the package level instance.

Someone could do `node.DefaultCopySettings.SomeField = true` and modify the package level variable.
But yes in our case, we always assign it to a variable, which copies it on the stack and since there is no pointer, it won't modify the package level variable this way.

> I don't see the need to call a function for this and allocating more memory.

Yes I agree, I changed it to global variable, a function call isn't really necessary here.
**However**, on the memory allocation side, TIL and also now you know:

```go
var globalVar = CopySettings{}

func getGlobalVar() CopySettings {
	return CopySettings{}
}

func BenchmarkXxx(b *testing.B) {
	b.Run("global variable", func(b *testing.B) {
		for i := 0; i < b.N; i++ {
			x := globalVar
			x.CopyCached = true
		}
	})

	b.Run("function", func(b *testing.B) {
		for i := 0; i < b.N; i++ {
			x := getGlobalVar()
			x.CopyCached = true
		}
	})
}
```

Gives

```
BenchmarkXxx/global_variable
BenchmarkXxx/global_variable-8         	1000000000	         0.3357 ns/op	       0 B/op	       0 allocs/op
BenchmarkXxx/function
BenchmarkXxx/function-8                	1000000000	         0.3395 ns/op	       0 B/op	       0 allocs/op
```

So both don't use any memory (I guess at least on the heap) and take really the same time.